proposals/TP2-dynamically-generated-cacheable-xds-resources.md

```diff
@@ -0,0 +1,722 @@
+message ResourceLocator {
+  // The resource name to subscribe to.
+  string name = 1;
+
+  // A set of constraints used to match against the dynamic parameters on the resource. This
+  // allows clients to select between multiple variants of the same resource.
+  DynamicParameterConstraints dynamic_parameter_constraints = 2;
+}
+```
+
+The following new field will be added to `DiscoveryRequest`, to allow clients
+to specify constraints when subscribing to a resource:
+
+```proto
+  // Alternative to resource_names field that allows specifying cache
+  // keys along with each resource name. If this is populated in the
+  // first request for a resource type on a stream, resource_names is ignored
+  // for all subsequent requests for that resource type on that stream.
+  // Clients that populate this field must be able to handle responses
+  // from the server where resources are wrapped in a Resource message.
+  repeated ResourceLocator resource_locators = 7;
+```
+
+Similarly, the following fields will be added to `DeltaDiscoveryRequest`:
+
+```proto
+  // Alternative to resource_names_subscribe field that allows specifying cache
+  // keys along with each resource name. If this is populated in the
+  // first request for a resource type on a stream, resource_names_subscribe
+  // and resource_names_unsubscribe are ignored for all subsequent requests
+  // for that resource type on that stream.
+  repeated ResourceLocator resource_locators_subscribe = 8;
+
+  // Alternative to resource_names_unsubscribe field that allows specifying cache
+  // keys along with each resource name. If resource_locators_subscribe is
+  // populated in the first request for a resource type on a stream,
+  // this field is used instead of resource_named_unsubscribe for all
+  // subsequent requests for that resource type on that stream.
+  repeated ResourceLocator resource_locators_unsubscribe = 9;
+```
+
+The following field will be added to the `Resource` message, to allow the
+server to return the dynamic parameters associated with each resource:
+
+```proto
+  // Dynamic parameters associated with this resource. To be used by client-side caches
+  // (including xDS proxies) when matching subscribed resource locators.
+  map<string, string> dynamic_parameters = 8;
+```
+
+### Migrating From Node Metadata
+
+Today, the equivalent of dynamic parameter constraints is node metadata,
+which can be used by servers to determine the set of resources to send
+for LDS and CDS wildcard subscriptions or to determine the contents of
+other resources (e.g., to select individual routes to be included in an
+RDS resource).  For transition purposes, this mechanism can continue
+to be supported by the client performing direct translation of node
+metadata to exact-match constraints.  For example, if the node metadata
+contains the entry `env=prod`, this would be translated to a constraint
+`{key_constraints:[{key:"env" value:{constraints:[{value:"prod"}]}}]}`.
```

Comment:
How do we know that `env` is from node metadata? I think this might matter in some cases, e.g. when the same key is used in node metadata and real DCP.